plrust/src/user_crate/ready.rs

```diff
@@ -27,13 +37,60 @@ impl FnReady {
         fn_oid: pg_sys::Oid,
         shared_object: Vec<u8>,
     ) -> eyre::Result<Self> {
-        // we write the shared object (`so`) bytes out to a temporary file rooted in our
-        // configured `plrust.work_dir`.  This will get removed from disk when this function
-        // exists, which is fine because we'll have dlopen()'d it by then and no longer need it
-        let temp_so_file = tempfile::Builder::new().tempfile_in(gucs::work_dir())?;
-        std::fs::write(&temp_so_file, shared_object)?;
+        #[cfg(target_os = "linux")]
+        let (file_holder, library) = {
+            // for Linux we write the `shared_object` bytes to a memory-mapped file of exactly the
+            // right size.  Then we ask `libloading::Library` to "dlopen" it using a direct path
+            // to its file descriptor in "/proc/self/fd/{raw_fd}".
+            //
+            // This is an added "safety" measure as we can (reasonably) assure ourselves that the
+            // file won't be overwritten between when we finish writing it and when it is dlopen'd
+            use std::io::Write;
+            use std::os::unix::io::AsRawFd;
+
+            let mfd = memfd::MemfdOptions::default()
+                .allow_sealing(true)
+                .create(&format!("plrust-fn-{db_oid}-{fn_oid}-{pg_proc_xmin}"))?;
+
+            // set the filesize to exactly what we know it should be
+            mfd.as_file().set_len(shared_object.len() as u64)?;
+
+            // make sure we can't change the filesize
+            mfd.add_seals(&[memfd::FileSeal::SealShrink, memfd::FileSeal::SealGrow])?;
+            mfd.add_seal(memfd::FileSeal::SealSeal)?;
+
+            // and write the shared_object bytes
+            mfd.as_file().write_all(&shared_object)?;
+
+            // generate a direct filename to the underlying raw file descriptor that `mfd` created
+            let raw_fd = mfd.as_raw_fd();
+            let filename = format!("/proc/self/fd/{raw_fd}");
+
+            // finally, load the library.  When this block ends the memory-mapped file will be
```

Comment:
Nit: memory-mapping is a different thing (you *can* memory-map anonymous files made with `memfd_create`, but they aren't always memory mapped).
```suggestion
            // finally, load the library.  When this block ends the anonymous file will be
```